scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/constraints/activities/ActivityExpression.java

```diff
@@ -614,4 +616,76 @@ public String prettyPrint(final String prefix) {
   @Override
   public void extractResources(final Set<String> names) { }
 
+  //true if second is a subset or equal to first
+  public static boolean subsetOrEqual(SerializedValue first, SerializedValue second){
+    final var visitor = new SerializedValue.Visitor<Boolean>(){
+      @Override
+      public Boolean onNull() {
+        return true;
+      }
+
+      @Override
+      public Boolean onNumeric(final BigDecimal value) {
+        final var firstAsMap = first.asNumeric();
+        if(firstAsMap.isEmpty()){
+          return false;
+        }
+        return firstAsMap.get().equals(value);
+      }
+
+      @Override
+      public Boolean onBoolean(final boolean value) {
+        final var firstAsBoolean = first.asBoolean();
+        if(firstAsBoolean.isEmpty()){
+          return false;
+        }
+        return firstAsBoolean.get().equals(value);
+      }
+
+      @Override
+      public Boolean onString(final String value) {
+        final var firstAsString = first.asString();
+        if(firstAsString.isEmpty()){
+          return false;
+        }
+        return firstAsString.get().equals(value);
+      }
+
+      @Override
+      public Boolean onMap(final Map<String, SerializedValue> value) {
+        final var firstAsMap = first.asMap();
+        if(firstAsMap.isEmpty()){
+          return false;
+        }
+        for(final var elementInSecond: value.entrySet()){
+          final var elementInFirst = firstAsMap.get().get(elementInSecond.getKey());
+          if(elementInFirst != null){
+            if(!subsetOrEqual(elementInFirst, elementInSecond.getValue())){
+              return false;
+            }
+          }
+        }
+        return true;
+      }
+
+      @Override
+      public Boolean onList(final List<SerializedValue> value) {
+        final var firstAsList = first.asList();
+        if(firstAsList.isEmpty()){
+          return false;
+        }
+        if(firstAsList.get().size() < value.size()){
+          return false;
+        }
+        for(int i = 0; i < value.size(); i++){
+          if(!subsetOrEqual(firstAsList.get().get(i), value.get(i))){
+            return false;
+          }
```

Comment:
I think we can consider that a list is an ordered data structure.

Reply:
If we did not wanted order, we would use a set ? A list is an ordered data structure. Then comparison should include order.

Reply:
This seems like another case of "syntactically equal" vs some other definition of match. Currently, all of our matches use syntactic equality - but there may be cases where users want a different definition of match (e.g. list contains, vs list equals exactly)

We are dabbling with this "looser" definition of match with this method `subsetOrEquals` - it expects strict equality on all types except for structs - for those it will match even if extra fields are present. That said, that's the same way typescript types work - a Typescript interface will match any object that has at least those fields. There's no way (that I'm aware of) to similarly define a type "sublist of X", or "contains same elements as X, independent of order". So, after considering the similarities with Typescript, I think it does make sense for our match to allow extra keys, but require strict equality on all other types

Reply:
Sure, although if we have to ensure that order equals, we need to be confident that neither of these Lists got their elements potentially reordered coming from Typescript/the DB. If we're confident in that, then I'm fine with resolving this.

Reply:
I think we can be confident. If such a reordering happens, the problem lies elsewhere.